### PR TITLE
Allow a custom manifest loader to provide custom environment variables

### PR DESCRIPTION
**Pull request by abertelrud**
Extend `ToolchainConfiguration` to allow a customization of the environment used when compiling package manifests.

### Motivation:

Clients of libSwiftPM can already provide a set of custom flags to pass to the Swift compiler when compiling and linking the manifest.  This extends this support to allow custom environment entries as well.  There's no immediate use for this in SwiftPM, but it could be used in the future, and there are clients that want to do so

### Modifications:

- add a new property to `ToolchainConfiguration` that contains a string-to-string mapping
- default it to `ProcessEnv.vars` to maintain status quo — this happens to not be needed on Darwin but is needed on Linux
- use it when compiling manifests and plugin scripts
- add a unit tests that uses this functionality for a toy purpose (for testing)

A separate but related change (in a separate commit) is to make the properties of `ToolchainConfiguration` be mutable, so that clients can use `ToolchainConfiguration.default` and then mutate a single property without having to instantiate a new struct and pass all the properties from `ToolchainConfigratuion.default`.

### Result:

Clients of libSwiftPM can customize the environment used when compiling package manifests.

### CODE REVIEW

File: Sources/PackageLoading/ManifestLoader.swift [hunk 1]
```diff
@@ -614,8 +621,8 @@ public final class ManifestLoader: ManifestLoaderProtocol {
             stream <<< packageIdentity
             stream <<< manifestContents
             stream <<< toolsVersion.description
-            for key in env.keys.sorted(by: >) {
-                stream <<< key <<< env[key]! // forced unwrap safe
+            for (key, value) in env.sorted(by: { $0.key > $1.key }) {
+                stream <<< key <<< value
```
Comment (abertelrud) [inline Sources/PackageLoading/ManifestLoader.swift:625]: This is a separate commit for getting rid of a force unwrap that I was while making these changes.  I can split out into a separate PR if needed (might be best)

**Comment by abertelrud** [comment]
@swift-ci please smoke test

**Comment by abertelrud** [comment]
Marking as draft since https://github.com/apple/swift-package-manager/pull/3656 has changed a lot of what this is based on.

File: Sources/PackageLoading/ManifestLoader.swift (hunk 1 above)
Comment (tomerd) [reply Sources/PackageLoading/ManifestLoader.swift:625]: this is fine imo, small enough

**Comment by tomerd** [comment]
> Marking as draft until I can rework it, since #3656 has changed a lot of what this is based on.

lgtm, the rebase on top of #3656 should be easy I hope - nothing changed there that should materially impact this

**Comment by abertelrud** [comment]
> > Marking as draft until I can rework it, since #3656 has changed a lot of what this is based on.
> 
> lgtm, the rebase on top of #3656 should be easy I hope - nothing changed there that should materially impact this

Sound good — will rebase and mark as ready.

**Comment by abertelrud** [comment]
@swift-ci please smoke test

**Comment by abertelrud** [comment]
@swift-ci please smoke test

**Comment by abertelrud** [comment]
@swift-ci please smoke test

**Comment by abertelrud** [comment]
This is now ready to be merged once it has been reviewed.

**Review by tomerd — APPROVED** [review]

**Pull request merged**